modules/stochastic_tools/include/surrogates/GaussianProcessTrainer.h

```diff
@@ -0,0 +1,75 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "SurrogateTrainer.h"
+#include "CovarianceFunction.h"
+#include "Standardizer.h"
+#include <Eigen/Dense>
+
+#include "Distribution.h"
+
+class GaussianProcessTrainer : public SurrogateTrainer
+{
+public:
+  static InputParameters validParams();
+  GaussianProcessTrainer(const InputParameters & parameters);
+  virtual void initialSetup() override;
+  virtual void initialize() override;
+  virtual void execute() override;
+  virtual void finalize() override;
+
+protected:
+private:
+  /// Sampler from which the parameters were perturbed
+  Sampler * _sampler;
+
+  /// Vector postprocessor of the results from perturbing the model with _sampler
+  const VectorPostprocessorValue * _values_ptr = nullptr;
+
+  /// True when _sampler data is distributed
+  bool _values_distributed;
+
+  /// Total number of parameters/dimensions
+  unsigned int _n_params;
+
+  // Type of kernel function to be used. See CovarianceFunction for options
+  const MooseEnum & _kernel_type;
+
+  /// Paramaters (x) used for training, along with statistics
+  RealEigenMatrix & _training_params;
+
+  /// Standardizer for use with params (x)
+  StochasticTools::Standardizer & _param_standardizer;
+
+  /// Data (y) used for training
+  RealEigenMatrix & _training_data;
+
+  /// Standardizer for use with data (y)
+  StochasticTools::Standardizer & _data_standardizer;
+
+  /// An _n_sample by _n_sample covariance matrix constructed from the selected kernel function
+  RealEigenMatrix & _K;
+
+  /// A solve of Ax=b via Cholesky.
+  RealEigenMatrix & _K_results_solve;
+
+  /// Pointer to covariance function used for K matrix
+  std::unique_ptr<CovarianceFunction::CovarianceKernel> & _covar_function;
```

Comment:
The CovarianceKernel object has a default constructor, so the unqiue_ptr should not be needed. You can create the object and then populate the members.

Reply:
I think my issue (which may be due to poor design on my part), is that I never really want to construct a standard `CovarianceKernel` object. I want to construct either a `SquaredExponential` `Exponential` or `MaternHalfInt` object. (each of these difference kernels would also have different constructor arguments if I make the other suggested change to use references for the data, which is totally valid)

In a normal workflow this wouldn't be much of an issue, because I would just conditionally construct the type of kernel I need.

However, because I need to send the kernel from the trainer to the surrogate, I need to use dataLoad of some form. In the current design, I am choosing to load the entire kernel function object. The problem lies in that (from what I can tell) the framework the dataLoad is called from creates the object before dataLoad is called, and dataLoad's job is to set that attributes to clone to what was saved. However the framework has no idea what type of object to create (`SquaredExponential` `Exponential` or `MaternHalfInt` ) which is why I think I opted for the pointer. Calling `_covar_function(getModelData<CovarianceFunction::CovarianceKernel>("_covar_function"))` would just construct a CovarianceKernel object correct? which I don't necessarily want to do, I want one of the other objects that derives form CovarianceKernel.

I also don't know how this reconciles with the other suggested change by Zach, in which the parameters would be set in the constructor (no more default constructor, unless I retain use of the set functions in the dataLoad)


At least 2-ish other designs would work around this, and each has their drawbacks. Keep in mind that there are MANY other CovarianceKernels that could/will be added, and each one comes with its own set of hyperparameters beyond the ones listed in CovairanceFunction.h (_sigma_f_squared, _sigma_n_squared, _length_factor, _p, _gamma). SO what ever solution is chosen should be easily expandable. 

### 
**Alternate option 1**
dataStore/dataLoad just operates on the attributes, saving and loading (_sigma_f_squared, _sigma_n_squared, _length_factor, _p, _gamma), all separately. We then conditionally construct the kernel type we want in the Surrogate.
```
if (type=="squared_exponential")
       kernel=SquaredExponential(_sigma_f_squared, _sigma_n_squared, _length_factor );
'if (type=="exponential")
       kernel=Exponential(_sigma_f_squared, _sigma_n_squared, _length_factor, _gamma );
'if (type=="matern_half_int")
       kernel=MaternHalfInt(_sigma_f_squared, _sigma_n_squared, _length_factor_, _p);
```
This keys each kernel type it's own distinct class and I believe the base CovarianceKernel becomes a true pure virtual, removes the issue of the dataStore/dataLoad, and gets rid of the pointer.

However as more an more kernels are added, the surrogate builds up a spaghetti `if (type==) `statements, and the number of dataStore/dataLoads grows as we need to load more and more different hyperparameters (maybe regardless of the actual selected kernel types uses them)

**Option 1.B**
Same as option 1 above, but pack the hyperparameters into one object to be stored and loaded (such as a std::vector). This is then parsed out somewhere based on the selected kernel type and used to create the kernel. Fixes the buildup of dataStore/Load statements by adding more kernels, but the spaghetti `if (type==) `statements remain, and MAY have to deal with the parsing as well.

**Option 2**
We forget about the one unique class per kernel function type, and the all exist as a single CovarianceFunction class, which has a `type` attribute, and we fill up the 'Compute_K' member function with `if (type==) `statements. The Covariance Function class gets messy but is mostly self contained to that.

When the hyperparameter tuning dev comes, we also need to comptue a bunch of Compute_dKd{param}, so these will also need the `if (type==) `statements.




Overall I agree that maybe the current implementation is maybe not the best way to do it, and maybe we need a small design meeting.

Reply:
If the resulting object has consistent interface with respect to the trainer (i.e., it always calls computeK) then the separate classes are good. Which I think is the case here.

Is it possible that these options should just each be a trainer? The type of trainer ("exponential", etc.) could be stored and the model can then create the correct helper class based on that? You can create the base GausiannProcessBase that does most of the work and the children just implement these various versions.

Sorry, if I am confusing the situation. I am still trying to grasp what this is doing.

Reply:
Having a separate trainer for each type of covariance function could be a good option. Maybe we can go a step further and create new type of MooseObject for the covariance. So the input could look like this:
```
[GaussianCovariance]
  [exponential]
    type = Exponential
    parameter1 = foo
    parameter2 = bar
  []
[]

[Trainers]
  [gp]
    type = GaussainProcessTrainer
    covariance_function = exponential
  []
[]
```

Just another idea to throw out there.

Reply:
Here is another idea:

```
[Trainers]
  [gp]
    type = GaussianProcessTrainer
    [GaussianCovariance]
     type = Exponential
   []
[]
[]
```

Reply:
> If the resulting object has consistent interface with respect to the trainer (i.e., it always calls computeK) then the separate classes are good. Which I think is the case here.

This is the case, and as I think about it is probably the way we should go **if** we do not want to create any new moose objects. The current implementation of CovarianceKernels seems to already %80 of the way there. 

> Is it possible that these options should just each be a trainer? The type of trainer ("exponential", etc.) could be stored and the model can then create the correct helper class based on that? You can create the base GausiannProcessBase that does most of the work and the children just implement these various versions.

I see what you mean here. MAYBE from a certain perspective it strays a bit too far in treating two different GP's with differing kernel functions as two very different things? Given that this would create new c;asses and MooseObjects for each covariance function implemented, I sort of tend towards the other options. (Which also add more classes and/or MooseObjects, but keep all this under the "realm" of a single trainer).

Regarding the two suggestions of having a separate MooseObject for covariance functions: I strongly prefer Andrews syntax. I haven't played around with this type of MooseObject, is it implemented just like any other, but there is a flag that it can ONLY be used in a certain context [trainers/gp]? Question on Andrew's idea though
```
[Trainers]
  [gp]
    type = GaussianProcessTrainer
    [GaussianCovariance]
     type = Exponential
   []
[]
[]
```
vs
 
```
[Trainers]
  [gp]
    type = GaussianProcessTrainer
    [Exponential]
     param_1 = something
   []
[]
[]
```
In Andrew's framework can we have the ability to have each type it's own moose object? (how Zach had it)


So yeah, at the moment I am either leaning towards separate classes for each covariance kernel, or a MooseObject based approach for the kernels in which they are constrained to the GP for use.

Reply:
The sub-block must have a "type =", so the bottom method is not an option.

In code there isn't really a difference between the syntax proposed by either me or @zachmprince. Each Covariance calculation will be a separate MooseObject. The difference is just in the input file syntax and the supporting action(s).

Reply:
Ahh sorry, I was thinking about `type` as a standard parameter I would have to deal with manually, and I forgot MOOSE deals with that specially.

Reply:
OK, this went through a a couple revisions and is now working. Utilizes the new CovarianceFunctionBase derived classes to specify the covariance function. 

When the Trainer and Surrogate are in the same input file, the surrogate gets the covariance from the Trainer using the SurrogateModelInterface 

Then the Surrogate is loading from a file, the LoadCovarianceDataAction reads hyper-parameters stored in map(s) in the training file, and uses those to reconstruct a covariance function object for the Surrogate to use.